juju/charm.py

```diff
@@ -30,3 +52,24 @@ class CharmEvents(EventsBase):
 class CharmBase(Object):
 
     on = CharmEvents()
+
+    def __init__(self, metadata, framework, key):
+        super().__init__(framework, key)
+        self.metadata = metadata
+
+        for role in ('requires', 'provides', 'peers'):
+            for relation_name in metadata.get(role, {}).keys():
```

Comment:
Not a huge deal given this is done once per execution, but this is both simpler and faster in those cases (no throw-away dicts allocated):
```
for relation_name in metadata.get(role, ()):
```